### PR TITLE
Precompute flash attention padding info

### PR DESCRIPTION
**Pull request by ShashankMosaicML**
This PR precomputes the padding information for flash attention once and reuses that for each layer, thus saving recomputation.

### CODE REVIEW

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 1]
```diff
@@ -216,6 +224,45 @@ def gen_attention_mask_in_length(sequence_id: Union[None, torch.Tensor], S: int,
...
+        query_padding_mask = attention_mask_in_length
+        unpadding_function = bert_padding.unpad_input_for_concatenated_sequences
+
+    _, indices_q, cu_seqlens_q, max_seqlen_q = unpadding_function(
+        torch.zeros(bsz, S, 1).to(device=device), query_padding_mask)
+    _, indices_k, cu_seqlens_k, max_seqlen_k = unpadding_function(
+        torch.zeros(bsz, past_key_len + S, 1).to(device=device),
+        key_padding_mask)
+    _, indices_v, _, _ = unpadding_function(
+        torch.zeros(bsz, past_key_len + S, 1).to(device=device),
```
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:253]:
```suggestion
            key_padding_mask = torch.ones((bsz, past_key_len + S),
                                          dtype=torch.bool, device=device)
        query_padding_mask = key_padding_mask[:, -S:]
        unpadding_function = bert_padding.unpad_input
    else:
        key_padding_mask = attention_mask_in_length
        query_padding_mask = attention_mask_in_length
        unpadding_function = bert_padding.unpad_input_for_concatenated_sequences

    _, indices_q, cu_seqlens_q, max_seqlen_q = unpadding_function(
        torch.zeros(bsz, S, 1, device=device), query_padding_mask)
    _, indices_k, cu_seqlens_k, max_seqlen_k = unpadding_function(
        torch.zeros(bsz, past_key_len + S, 1, device=device),
        key_padding_mask)
    _, indices_v, _, _ = unpadding_function(
        torch.zeros(bsz, past_key_len + S, 1, device=device),
```

instantiate on GPU instead of on CPU then move to GPU

**Review by vchiley — APPROVED** [review]

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 2]
```diff
@@ -216,6 +224,44 @@ def gen_attention_mask_in_length(sequence_id: Union[None, torch.Tensor], S: int,
...
+        key_padding_mask = attention_mask_in_length
+        query_padding_mask = attention_mask_in_length
+        unpadding_function = bert_padding.unpad_input_for_concatenated_sequences
+
+    _, indices_q, cu_seqlens_q, max_seqlen_q = unpadding_function(
+        torch.zeros(bsz, S, 1, device=device), query_padding_mask)
+    _, indices_k, cu_seqlens_k, max_seqlen_k = unpadding_function(
+        torch.zeros(bsz, past_key_len + S, 1, device=device), key_padding_mask)
+    _, indices_v, _, _ = unpadding_function(
+        torch.zeros(bsz, past_key_len + S, 1, device=device), key_padding_mask)
```
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:253]:
```suggestion
        torch.empty(bsz, S, 1, device=device), query_padding_mask)
    _, indices_k, cu_seqlens_k, max_seqlen_k = unpadding_function(
        torch.empty(bsz, past_key_len + S, 1, device=device), key_padding_mask)
    _, indices_v, _, _ = unpadding_function(
        torch.empty(bsz, past_key_len + S, 1, device=device), key_padding_mask)
```

since its dummy data, there is probably no need for it to be initialized

**Review by mvpatel2000 — APPROVED** [review]

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 3]
```diff
@@ -24,15 +24,23 @@
...
         from flash_attn.layers.rotary import \
             RotaryEmbedding as DAILRotaryEmbedding
     except Exception as e:
         raise e
 
+if is_flash_v1_installed():
+    try:  # This try...except is needed because transformers requires it despite the 'if' statement above
+        from flash_attn import bert_padding
+    except Exception as e:
+        raise e
```
Comment (mvpatel2000) [inline llmfoundry/models/mpt/modeling_mpt.py:42]: Why is this function both defined and imported...

**Review by dakinggg — APPROVED** [review]

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 3 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/modeling_mpt.py:42]: That's an `if` condition :D

File: llmfoundry/models/layers/attention.py [hunk 4]
```diff
@@ -228,12 +228,15 @@ def flash_attn_fn(
...
     needs_weights: bool = False,
     multiquery: bool = False,
-    attention_mask_in_length: Optional[torch.Tensor] = None,
     should_repeat_kv_for_gqa: Optional[bool] = True,
     sliding_window_size: int = -1,
     alibi_slopes: Optional[torch.Tensor] = None,
+    flash_attn_padding_info: Optional[dict[str, torch.Tensor]] = None,
 ) -> tuple[torch.Tensor, Optional[torch.Tensor], Optional[tuple[torch.Tensor,
                                                                 torch.Tensor]]]:
+    del key_padding_mask
```
[the remark this answers is not in the document]
Comment (ShashankMosaicML) [reply llmfoundry/models/layers/attention.py:239]: Done.
